receiver/hostmetricsreceiver/internal/scraper/cpuscraper/internal/metadata/generated_metrics.go

```diff
@@ -17,81 +17,180 @@
+type metricBuilder struct {
+	metadata        metricMetadata
+	config          MetricConfig
+	metric          pdata.Metric
+	initialCapacity int
+	startTime       pdata.Timestamp
 }
 
-// Name returns the metric name.
-func (m *metricImpl) Name() string {
-	return m.name
+// MetricBuilderOption applies changes to default metrics builder.
+type MetricBuilderOption func(metricBuilder)
+
+// WithStartTime sets startTime on the metrics builder.
+func WithStartTime(startTime pdata.Timestamp) MetricBuilderOption {
+	return func(mb metricBuilder) {
+		mb.startTime = startTime
+	}
 }
 
-// New creates a metric object preinitialized.
-func (m *metricImpl) New() pdata.Metric {
-	metric := pdata.NewMetric()
-	m.Init(metric)
-	return metric
+// WithInitialCapacity sets initial capacity for metric data points.
+func WithInitialCapacity(initialCapacity int) MetricBuilderOption {
+	return func(mb metricBuilder) {
+		mb.initialCapacity = initialCapacity
+	}
 }
 
-// Init initializes the provided metric object.
-func (m *metricImpl) Init(metric pdata.Metric) {
-	m.initFunc(metric)
+func newMetricBuilder(metadata metricMetadata, config MetricConfig, options ...MetricBuilderOption) metricBuilder {
+	mb := metricBuilder{
+		metadata: metadata,
+		config:   config,
+	}
+	if !mb.Enabled() {
+		return mb
+	}
+
+	mb.createMetric()
+	mb.startTime = pdata.NewTimestampFromTime(time.Now())
+
+	for _, op := range options {
+		op(mb)
+	}
+	return mb
 }
 
-type metricStruct struct {
-	SystemCPUTime MetricIntf
+// Enabled identifies whether the metrics should be collected or not.
+func (mb *metricBuilder) Enabled() bool {
+	if mb.config.Enabled != nil {
+		return *mb.config.Enabled
+	}
+	return mb.metadata.enabled
+}
+
+// EnsureDataPointsCapacity ensures metric data points slice capacity.
+func (mb *metricBuilder) EnsureDataPointsCapacity(cap int) {
+	if !mb.Enabled() {
+		return
+	}
+	mb.metric.Sum().DataPoints().EnsureCapacity(cap)
+}
+
+// Reset resets the metric builder startTime and removes previous/current metric state.
+func (mb metricBuilder) Reset(options ...MetricBuilderOption) {
+	if !mb.Enabled() {
+		return
+	}
+	mb.startTime = pdata.NewTimestampFromTime(time.Now())
+	for _, op := range options {
+		op(mb)
+	}
+	mb.createMetric()
 }
 
-// Names returns a list of all the metric name strings.
-func (m *metricStruct) Names() []string {
-	return []string{
-		"system.cpu.time",
+// Collect appends generated metric to a pdata.MetricsSlice and updates the internal state to be ready for recording
+// another set of data points. This function will be doing all transformation required to produce metric representation
+// defined in metadata and user configuration, e.g. delta/cumulative translation.
+func (mb *metricBuilder) Collect(metrics pdata.MetricSlice) {
+	if !mb.Enabled() {
+		return
 	}
+
+	mb.metric.CopyTo(metrics.AppendEmpty())
+
+	// Reset metric data points collection.
+	mb.createMetric()
```

Comment:
mb.DataPoints().Clear()? Since we copy everything anyway. Can be optimized later.

Reply:
I don't think we have Clear method for data points, only for attributes AFAIK. Should we add one for this use case?

Reply:
Not now, just a TODO is enough for the moment, we can optimize after we have something working.